### DISCUSSION TITLE
Specifing input variables for Linearization in Physics/Mechanics

### DISCUSSION THREAD

**Discussion by AlessandroZavoli** [sympy/sympy#26199] [Q&A] [answered]
When using the linearization routines in [sympy.physics.mechanics](https://docs.sympy.org/latest/modules/physics/mechanics/linearize.html)
it is not clear how to define the system inputs $r$

In particular, the form
$M \delta x' =  A \delta x + B \delta r$
returns B empty even in the case of applied non-conservative forces that depend on a parameter (i.e., the system input).

What is the correct way to specify that a symbolic variable is a system input?

**Comment by moorepants** [comment, marked as answer]
I don't think you have to do anything. If you have symbols that are functions of time in your equations of motion that are not state variables, then the linearize method treats them as inputs.

**Reply by AlessandroZavoli** [reply]
Do you mean that r is needs to  be a Function of t or a dynamicvariable?

**Reply by moorepants** [reply]
Yes.

**Comment by AlessandroZavoli** [comment]
Using a dynamicsimbol, the matrix B is no longer empty.
Yet, I think something is not working properly: before the linearization i would expect to be able to set the operating-point value (so that the EoMs are respected).
Alos, I should be able to retrieve the list of reference inputs in some way after the linearization

**Reply by moorepants** [reply]
You can set the operating point for any dynamicsymbol in that kwarg. The method returns A, B, and r, no? or M, A, B, r. So the list of inputs is there.

**Reply by AlessandroZavoli** [reply]
Using the dynamicsymbols the matrix B is now non-empty. Yet, the method does not return r as 4th argument 
`M, A, B, r = linearizer.linearize(A_and_B=False, op_point=op_point)` returns `ValueError: not enough values to unpack (expected 4, got 3)`.
Nevertheless, by the reference variables are correctly stored in `linearizer.r` on the creation of the linearizer object.

I will try to write a minimum example to reproduce the error (or point me to one if you have it)
Best, 
Alessandro

**Reply by moorepants** [reply]
It doesn't look like the `Linearizer` returns `r`:

https://github.com/sympy/sympy/blob/master/sympy/physics/mechanics/linearize.py#L259

**Reply by moorepants** [reply]
The `linearize()` methods of LagrangesMethod and KanesMethod return `r`: https://github.com/sympy/sympy/blob/master/sympy/physics/mechanics/kane.py#L632
